### PR TITLE
PT-7938: add addOrUpdateOrderPayment mutation

### PR DESCRIPTION
**Pull request by ksavosteev**
## Description
```js
mutation addOrUpdateOrderPayment($command: InputAddOrUpdateOrderPaymentType!) {
addOrUpdateOrderPayment(command: $command) {
    	id
      inPayments {
        id
        createdDate
          billingAddress {
          id
          city
          countryCode
          countryName
          email
          firstName
          lastName
          line1
          line2
          middleName
          name
          organization
          phone
          postalCode
          regionId
          regionName
          zip
          addressType
        }
        paymentMethod {
          code
          paymentMethodType
          paymentMethodGroupType
        }
    	}
    }
  }

```
---
```json
{
    "command": {
        "orderId": "74d8b492-0bb5-486e-a0e6-0915848a7379",
        "payment": {
            "id": "123",
            "paymentGatewayCode": "AuthorizeNetPaymentMethod",          
            "billingAddress": {
                "city": "Test",
                "countryCode": "US",
                "countryName": "US",
                "email": "address@mail.test",
                "firstName": "First test name",
                "id": "KeyTest",
                "key": "KeyTest",
                "lastName": "Last name test",
                "line1": "Address Line 1",
                "line2": "Address line 2",
                "middleName": "Test Middle Name",
                "name": "First name address",
                "organization": "OrganizationTestName",
                "phone": "88005553535",
                "postalCode": "111111",
                "regionId": "Test region",
                "regionName": "Region 15",
                "zip": "13413",
                "addressType": 2
            }
        }
    }
}
```
### Jira-link:

https://virtocommerce.atlassian.net/browse/PT-7938

### CODE REVIEW

File: src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Commands/AddOrUpdateOrderPaymentCommandHandler.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+using VirtoCommerce.Platform.Core.Common;
+using VirtoCommerce.Platform.Core.GenericCrud;
+
+namespace VirtoCommerce.ExperienceApiModule.XOrder.Commands
+{
+    public class AddOrUpdateOrderPaymentCommandHandler : IRequestHandler<AddOrUpdateOrderPaymentCommand, CustomerOrderAggregate>
+    {
+        private readonly ICustomerOrderAggregateRepository _customerOrderAggregateRepository;
+        private readonly ICustomerOrderService _customerOrderService;
+        private readonly ISearchService<PaymentMethodsSearchCriteria, PaymentMethodsSearchResult, PaymentMethod> _paymentMethodsSearchService;
```
Comment (Vectorfield4) [inline src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Commands/AddOrUpdateOrderPaymentCommandHandler.cs:19]: Can you inject `IPaymentSearchService`?

File: src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Commands/AddOrUpdateOrderPaymentCommandHandler.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            _customerOrderAggregateRepository = customerOrderAggregateRepository;
+            _customerOrderService = customerOrderService;
+            _paymentMethodsSearchService = (ISearchService<PaymentMethodsSearchCriteria, PaymentMethodsSearchResult, PaymentMethod>)paymentMethodsSearchService;
+        }
+
+        public async Task<CustomerOrderAggregate> Handle(AddOrUpdateOrderPaymentCommand request, CancellationToken cancellationToken)
+        {
+            var orderAggregate = await _customerOrderAggregateRepository.GetOrderByIdAsync(request.OrderId);
+
+            var paymentId = request.Payment.Id?.Value ?? null;
```
Comment (Vectorfield4) [inline src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Commands/AddOrUpdateOrderPaymentCommandHandler.cs:35]: Redundant operator
```suggestion
            var paymentId = request.Payment.Id?.Value;
```

File: src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Commands/AddOrUpdateOrderPaymentCommandHandler.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            var payment = orderAggregate.Order.InPayments.FirstOrDefault(s => paymentId != null && s.Id == paymentId);
+            payment = request.Payment.MapTo(payment);
```
Comment (Vectorfield4) [inline src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Commands/AddOrUpdateOrderPaymentCommandHandler.cs:37]: Should we check payment for null before `MapTo(payment)` call?

File: src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/CustomerOrderAggregate.cs [hunk 4]
```diff
@@ -123,6 +125,52 @@ public virtual async Task<CustomerOrderAggregate> UpdatePaymentDynamicProperties
             return this;
         }
 
+        public virtual async Task<CustomerOrderAggregate> AddPaymentAsync(PaymentIn payment, IEnumerable<PaymentMethod> availPaymentMethods)
```
Comment (Vectorfield4) [inline src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/CustomerOrderAggregate.cs:128]:
```suggestion
        public virtual async Task<CustomerOrderAggregate> AddPaymentAsync([DisallowNull] PaymentIn payment, IEnumerable<PaymentMethod> availPaymentMethods)
```

File: src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/CustomerOrderAggregate.cs [hunk 5]
```diff
@@ -123,6 +125,52 @@ public virtual async Task<CustomerOrderAggregate> UpdatePaymentDynamicProperties
...
+
+            if (payment?.Currency == null)
+            {
+                payment.Currency = Order.Currency;
+            }
+
+            if (payment.CustomerId == null)
+            {
+                payment.CustomerId = Order.CustomerId;
+            }
```
Comment (Vectorfield4) [inline src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/CustomerOrderAggregate.cs:146]:
```suggestion
            payment.Currency ??= Order.Currency;
            payment.CustomerId ??= Order.CustomerId;
```

File: src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/CustomerOrderAggregate.cs [hunk 6]
```diff
@@ -123,6 +125,52 @@ public virtual async Task<CustomerOrderAggregate> UpdatePaymentDynamicProperties
...
+            {
+                var existingPayment = !payment.IsTransient() ? Order.InPayments.FirstOrDefault(s => s.Id == payment.Id) : null;
+                if (existingPayment != null)
+                {
+                    Order.InPayments.Remove(existingPayment);
+                }
+            }
+
+            return Task.FromResult(this);
+        }
```
Comment (Vectorfield4) [inline src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/CustomerOrderAggregate.cs:172]:
```suggestion
        protected virtual Task<CustomerOrderAggregate> RemoveExistingPaymentAsync([DisallowNull] PaymentIn payment)
        {
            var existingPayment = !payment.IsTransient() ? Order.InPayments.FirstOrDefault(s => s.Id == payment.Id) : null;
            if (existingPayment != null)
            {
                Order.InPayments.Remove(existingPayment);
            }

            return Task.FromResult(this);
        }
```

File: src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Models/ExpOrderAddress.cs [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        public Optional<string> Zip { get; set; }
+        public Optional<string> OuterId { get; set; }
+        public Optional<int> AddressType { get; set; }
+
+        public Address MapTo(Address address)
+        {
+            if (address == null)
+            {
+                address = AbstractTypeFactory<Address>.TryCreateInstance();
+            }
```
Comment (Vectorfield4) [inline src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Models/ExpOrderAddress.cs:35]:
```suggestion
            address ??= AbstractTypeFactory<Address>.TryCreateInstance();
```

File: src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Models/ExpOrderAddress.cs [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+            if (OuterId?.IsSpecified == true)
+            {
+                address.OuterId = OuterId.Value;
+            }
+
+            if (AddressType?.IsSpecified == true)
+            {
+                address.AddressType = (AddressType)AddressType.Value;
+            }
```
Comment (Vectorfield4) [inline src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Models/ExpOrderAddress.cs:130]: Boilerplate code. Maybe add some reflection?
@VirtoCommerce/backend-developers

File: src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Models/ExpOrderPayment.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        public Optional<string> PaymentGatewayCode { get; set; }
+        public Optional<string> Currency { get; set; }
+        public Optional<decimal> Price { get; set; }
+        public Optional<decimal> Amount { get; set; }
+
+        public Optional<ExpOrderAddress> BillingAddress { get; set; }
+
+        public IList<DynamicPropertyValue> DynamicProperties { get; set; }
+
+        public PaymentIn MapTo(PaymentIn payment)
```
Comment (Vectorfield4) [inline src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Models/ExpOrderPayment.cs:22]: Same.

File: src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/OrderErrorDescriber.cs [hunk 10]
```diff
@@ -33,5 +33,10 @@ public static string StoreNotFound()
         {
             return $"Can't find a store";
         }
+
+        public static string PaymentMethodUnavailable(string code)
+        {
+            return $"The payment method code:{code} unavailable";
+        }
```
Comment (Vectorfield4) [inline src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/OrderErrorDescriber.cs:40]: Move string format to constant.

File: src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Schemas/InputPaymentInType.cs [hunk 11]
```diff
@@ -1,8 +1,10 @@
+using System;
 using GraphQL.Types;
 using VirtoCommerce.OrdersModule.Core.Model;
 
 namespace VirtoCommerce.ExperienceApiModule.XOrder.Schemas
 {
+    [Obsolete("Use InputOrderPaymentType.")]
```
Comment (Vectorfield4) [inline src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Schemas/InputPaymentInType.cs:7]:
```suggestion
    [Obsolete("Use InputOrderPaymentType. Should be removed in stable v3")]
```

**Review by Vectorfield4 — APPROVED** [review]

File: src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Commands/AddOrUpdateOrderPaymentCommandHandler.cs (hunk 1 above)
Comment (ksavosteev) [reply src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Commands/AddOrUpdateOrderPaymentCommandHandler.cs:19]: already did

File: src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Commands/AddOrUpdateOrderPaymentCommandHandler.cs (hunk 3 above)
Comment (ksavosteev) [reply src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Commands/AddOrUpdateOrderPaymentCommandHandler.cs:37]: payment  will be created in MapTo if null

File: src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/OrderErrorDescriber.cs (hunk 10 above)
Comment (ksavosteev) [reply src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/OrderErrorDescriber.cs:40]: can't use string interpolation then

**Pull request merged**